File: services/localDb.service.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DEFAULT_DB_PATH = path.join(__dirname, "../data/processed/database.json");

function safeLower(s) {
  return (s || "").toString().trim().toLowerCase();
}

function getMaxNumericId(items, key) {
  let max = 0;
  for (const item of items || []) {
    const raw = item?.[key];
    const n = typeof raw === "number" ? raw : parseInt(raw, 10);
    if (Number.isFinite(n) && n > max) max = n;
  }
  return max;
}

export function loadLocalDb(dbPath = DEFAULT_DB_PATH) {
  if (!fs.existsSync(dbPath)) return [];
  const raw = fs.readFileSync(dbPath, "utf-8");
  const parsed = JSON.parse(raw);
  return Array.isArray(parsed) ? parsed : [];
}

export function saveLocalDb(db, dbPath = DEFAULT_DB_PATH) {
  const dir = path.dirname(dbPath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(dbPath, JSON.stringify(db, null, 2), "utf-8");
}

export function findDestination(db, destinationName) {
  const d = safeLower(destinationName);
  if (!d) return null;

  // Prefer exact match; fallback to substring match.
  const exact = db.find(p => safeLower(p?.place_name) === d);
  if (exact) return exact;
  return db.find(p => safeLower(p?.place_name).includes(d));
}

export function upsertDestination(db, destination) {
  const nextDb = Array.isArray(db) ? db : [];
  const nameKey = safeLower(destination?.place_name);
  if (!nameKey) return { db: nextDb, destination: null, created: false, updated: false };

  const idx = nextDb.findIndex(p => safeLower(p?.place_name) === nameKey);
  if (idx >= 0) {
    const existing = nextDb[idx] || {};
    // Merge but keep existing nested arrays if present.
    const merged = {
      ...existing,
      ...destination,
      attractions: Array.isArray(destination?.attractions)
        ? destination.attractions
        : (existing.attractions || [])
    };
    nextDb[idx] = merged;
    return { db: nextDb, destination: merged, created: false, updated: true };
  }

  // Assign new place_id if missing.
  const maxPlaceId = getMaxNumericId(nextDb, "place_id");
  const toInsert = {
    place_id: destination?.place_id ?? String(maxPlaceId + 1),
    state: destination?.state ?? "Unknown",
    description: destination?.description ?? "",
    attractions: Array.isArray(destination?.attractions) ? destination.attractions : [],
    ...destination
  };
  nextDb.push(toInsert);
  return { db: nextDb, destination: toInsert, created: true, updated: false };
}

export function buildDestinationFromApi({ destinationName, coords, places = [] }, existingDb = []) {
  const maxPlaceId = getMaxNumericId(existingDb, "place_id");
  const maxSpotId = (() => {
    let max = 0;
    for (const dest of existingDb) {
      const attractions = dest?.attractions || [];
      const m = getMaxNumericId(attractions, "spot_id");
      if (m > max) max = m;
    }
    return max;
  })();

  const placeId = String(maxPlaceId + 1);

  const attractions = (places || [])
    .filter(p => p && (p.name || p.place_name))
    .slice(0, 20)
    .map((p, i) => {
      const spotId = String(maxSpotId + 1 + i);
      const spotName = p.name || p.place_name;
      return {
        spot_id: spotId,
        place_id: placeId,
        spot_name: spotName,
        description: p.address || p.category || p.type || "Attraction",
        lat: p.lat ?? null,
        lon: p.lon ?? null,
        dining: [],
        accommodation: []
      };
    });

  return {
    place_id: placeId,
    place_name: destinationName,
    lat: coords?.lat ?? null,
    lon: coords?.lon ?? null,
    state: "Unknown",
    description: "Auto-added from live APIs",
    attractions
  };
}

File: services/local_db_service.py
import os
import json
import logging

logger = logging.getLogger(__name__)

DEFAULT_DB_PATH = os.path.join(os.path.dirname(__file__), "../data/processed/database.json")

def load_local_db(db_path=DEFAULT_DB_PATH):
    if not os.path.exists(db_path):
        return []
    try:
        with open(db_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
            return data if isinstance(data, list) else []
    except Exception as e:
        logger.error(f"Failed to load DB: {e}")
        return []

def save_local_db(db, db_path=DEFAULT_DB_PATH):
    try:
        os.makedirs(os.path.dirname(db_path), exist_ok=True)
        with open(db_path, 'w', encoding='utf-8') as f:
            json.dump(db, f, indent=2, ensure_ascii=False)
    except Exception as e:
        logger.error(f"Failed to save DB: {e}")

def safe_lower(s):
    return str(s).strip().lower() if s else ""

def find_destination(db, destination_name):
    d = safe_lower(destination_name)
    if not d: return None
    
    # Exact match
    for p in db:
        if safe_lower(p.get("place_name")) == d:
            return p
            
    # Substring match
    for p in db:
        if d in safe_lower(p.get("place_name")):
            return p
            
    return None

def get_max_numeric_id(items, key):
    max_val = 0
    for item in items or []:
        raw = item.get(key)
        try:
            val = int(raw)
            if val > max_val:
                max_val = val
        except (ValueError, TypeError):
            continue
    return max_val

def upsert_destination(db, destination):
    if not isinstance(db, list): db = []
    
    name_key = safe_lower(destination.get("place_name"))
    if not name_key:
        return {"db": db, "destination": None, "created": False, "updated": False}

    idx = -1
    for i, p in enumerate(db):
        if safe_lower(p.get("place_name")) == name_key:
            idx = i
            break
            
    if idx >= 0:
        existing = db[idx]
        merged = existing.copy()
        merged.update(destination)
        # Preserve existing attractions if not overwritten
        if "attractions" not in destination:
             merged["attractions"] = existing.get("attractions", [])
             
        db[idx] = merged
        return {"db": db, "destination": merged, "created": False, "updated": True}

    # Insert new
    max_place_id = get_max_numeric_id(db, "place_id")
    destination["place_id"] = destination.get("place_id", str(max_place_id + 1))
    destination["state"] = destination.get("state", "Unknown")
    destination["description"] = destination.get("description", "")
    destination["attractions"] = destination.get("attractions", [])
    
    db.append(destination)
    return {"db": db, "destination": destination, "created": True, "updated": False}

def build_destination_from_api(data, existing_db):
    destination_name = data.get("destinationName")
    coords = data.get("coords")
    places = data.get("places", [])
    
    max_place_id = get_max_numeric_id(existing_db, "place_id")
    
    # Calculate max spot id
    max_spot_id = 0
    for dest in existing_db:
        for attr in dest.get("attractions", []):
            sid = get_max_numeric_id([attr], "spot_id")
            if sid > max_spot_id: max_spot_id = sid

    place_id = str(max_place_id + 1)
    
    attractions = []
    for i, p in enumerate(places[:20]):
        if not p: continue
        name = p.get("name") or p.get("place_name")
        if not name: continue
        
        attractions.append({
            "spot_id": str(max_spot_id + 1 + i),
            "place_id": place_id,
            "spot_name": name,
            "description": p.get("address") or p.get("category") or "Attraction",
            "lat": p.get("lat"),
            "lon": p.get("lon"),
            "dining": [],
            "accommodation": []
        })

    return {
        "place_id": place_id,
        "place_name": destination_name,
        "lat": coords.get("lat") if coords else None,
        "lon": coords.get("lon") if coords else None,
        "state": "Unknown",
        "description": "Auto-added from live APIs",
        "attractions": attractions
    }
